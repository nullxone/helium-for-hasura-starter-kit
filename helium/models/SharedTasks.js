const { belongsTo } = require("helium-for-hasura");

class SharedTasks {
  static get relationships() {
    return [
      belongsTo("with_user", require("./Users"), "with_user_id"),
      belongsTo("task", require("./Tasks"), "task_id"),
    ];
  }

  static get permissions() {
    const pManage = eq("task.user_id", "X-Hasura-User-Id"),
      pSelect = or(pManage, eq("with_user_id", "X-Hasura-User-Id")),
      cols = ["id", "with_user_id", "task_id"];
    return {
      select: permission(pSelect, cols),
      insert: permission(pManage, cols),
      update: permission(pManage, cols),
      delete: permission(pManage),
    };
  }
}

module.exports = SharedTasks;
