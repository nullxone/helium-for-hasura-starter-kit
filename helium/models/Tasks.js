const { belongsTo, permission } = require("helium-for-hasura");

class Tasks {
  static get relationships() {
    return [belongsTo("user", require("./Users"), "user_id")];
  }

  static get permissions() {
    const pred = eq("user_id", "X-Hasura-User-Id"),
      cols = ["id", "title", "body", "user_id"];
    return {
      select: permission(pred, cols),
      insert: permission(pred, cols),
      update: permission(pred, cols),
      delete: permission(pred),
    };
  }
}

module.exports = Tasks;
