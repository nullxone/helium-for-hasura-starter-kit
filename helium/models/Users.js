const { hasMany } = require("helium-for-hasura");

class Users {
  static get relationships() {
    return [hasMany("tasks", require("./Tasks"), "user_id")];
  }

  static get permissions() {
    const pred = eq("id", "X-Hasura-User-Id");
    return {
      update: permission(pred, ["name"]),
    };
  }
}

module.exports = Users;
