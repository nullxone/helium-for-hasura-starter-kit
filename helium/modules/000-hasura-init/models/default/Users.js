const { hasMany, hasOne, belongsTo, permission } = require("../index");

// prettier-ignore
class Users {
  static allColumns = ["id","name","created_at"];

  static get relationships() {
    return [
      
    ];
  }
}

module.exports = Users;
