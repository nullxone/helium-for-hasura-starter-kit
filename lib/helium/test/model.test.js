const { Model } = require("../model");
const path = require("path");
const fs = require("fs-extra");

test("class to tracked table", () => {
  const appDir = path.join(__dirname, "./AppSimple");
  fs.emptyDirSync(path.join(appDir, "metadata"));

  Model.compile(
    require(path.join(appDir, "helium/models/Users")),
    path.join(appDir, "metadata")
  );

  const outFile = path.join(
    appDir,
    "metadata/databases/default/tables/public_users.yaml"
  );

  expect(fs.existsSync(outFile)).toBe(true);
  expect(fs.readFileSync(outFile, "utf8")).toMatchSnapshot();
});

test("write all tables file", () => {
  const appDir = path.join(__dirname, "./AppSimple");
  const metadataDir = path.join(appDir, "metadata");
  fs.emptyDirSync(metadataDir);

  Model.compile(require(path.join(appDir, "helium/models/Users")), metadataDir);

  Model.compile(require(path.join(appDir, "helium/models/Tasks")), metadataDir);

  Model.writeAllTables(metadataDir);

  const outFile = path.join(
    metadataDir,
    "databases/default/tables/tables.yaml"
  );

  expect(fs.existsSync(outFile)).toBe(true);
  expect(fs.readFileSync(outFile, "utf8")).toMatchSnapshot();
});
