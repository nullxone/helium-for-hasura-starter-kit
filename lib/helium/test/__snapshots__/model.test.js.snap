// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`belongs to relationship, has many relationship 1`] = `
"table:
  name: projects
  schema: public
object_relationships:
  - name: user
    using:
      manual_configuration:
        column_mapping:
          user_id: id
        insertion_order: null
        remote_table:
          name: users
          schema: public
array_relationships:
  - name: tasks
    using:
      manual_configuration:
        column_mapping:
          id: project_id
        insertion_order: null
        remote_table:
          name: tasks
          schema: public
"
`;

exports[`class to tracked table, has one relationship 1`] = `
"table:
  name: users
  schema: public
object_relationships:
  - name: project
    using:
      manual_configuration:
        column_mapping:
          id: user_id
        insertion_order: null
        remote_table:
          name: projects
          schema: public
  - name: supervisor
    using:
      manual_configuration:
        column_mapping:
          supervisor_id: id
        insertion_order: null
        remote_table:
          name: users
          schema: public
"
`;

exports[`complex predicate 1`] = `
{
  "_and": [
    {
      "user": {
        "id": {
          "_eq": "X-Hasura-User-Id",
        },
      },
    },
    {
      "_or": [
        {
          "_and": [
            {
              "_not": {
                "is_reviewed": {
                  "_eq": true,
                },
              },
            },
            {
              "category": {
                "_eq": "editorial",
              },
            },
          ],
        },
        {
          "category": {
            "_ne": "editorial",
          },
        },
      ],
    },
    {
      "_exists": {
        "_table": {
          "name": "blog",
          "schema": "public",
        },
        "_where": {
          "authors": {
            "id": {
              "_eq": "X-Hasura-User-Id",
            },
          },
        },
      },
    },
  ],
}
`;

exports[`write all tables file 1`] = `
"- '!include public_tasks.yaml'
- '!include public_users.yaml'
"
`;
